examples/blake2f.rs

```diff
@@ -0,0 +1,1371 @@
+    },
+    plonkish::{
+        backend::halo2::{chiquitoSuperCircuit2Halo2, ChiquitoHalo2SuperCircuit},
+        compiler::{
+            cell_manager::{MaxWidthCellManager, SingleRowCellManager},
+            config,
+            step_selector::SimpleStepSelectorBuilder,
+        },
+        ir::sc::SuperCircuit,
+    },
+    poly::ToExpr,
+    sbpir::query::Queriable,
+};
+use halo2_proofs::{
+    dev::MockProver,
+    halo2curves::{bn256::Fr, group::ff::PrimeField},
+};
+use std::hash::Hash;
+
+pub const IV_LENGTH: usize = 8;
+pub const SIGMA_VECTOR_LENGTH: usize = 16;
+pub const R1: usize = 32;
+pub const R2: usize = 24;
+pub const R3: usize = 16;
+pub const R4: usize = 63;
+pub const MIXING_ROUND: usize = 12;
+pub const BITS_NUMBER: usize = 16;
+pub const VALUE_4BITS: u64 = 16;
+pub const XOR_4BITS_NUMBER: usize = BITS_NUMBER * BITS_NUMBER;
+pub const V_LEN: usize = 16;
+pub const M_LEN: usize = 16;
+pub const H_LEN: usize = 8;
+pub const G_ROUND: usize = 16;
+
+pub const IV_VALUES: [u64; IV_LENGTH] = [
+    0x6A09E667F3BCC908,
+    0xBB67AE8584CAA73B,
+    0x3C6EF372FE94F82B,
+    0xA54FF53A5F1D36F1,
+    0x510E527FADE682D1,
+    0x9B05688C2B3E6C1F,
+    0x1F83D9ABFB41BD6B,
+    0x5BE0CD19137E2179,
+];
+
+pub const SIGMA_VALUES: [[usize; SIGMA_VECTOR_LENGTH]; 10] = [
+    [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15],
+    [14, 10, 4, 8, 9, 15, 13, 6, 1, 12, 0, 2, 11, 7, 5, 3],
+    [11, 8, 12, 0, 5, 2, 15, 13, 10, 14, 3, 6, 7, 1, 9, 4],
+    [7, 9, 3, 1, 13, 12, 11, 14, 2, 6, 5, 10, 4, 0, 15, 8],
+    [9, 0, 5, 7, 2, 4, 10, 15, 14, 1, 11, 12, 6, 8, 3, 13],
+    [2, 12, 6, 10, 0, 11, 8, 3, 4, 13, 7, 5, 15, 14, 1, 9],
+    [12, 5, 1, 15, 14, 13, 4, 10, 0, 7, 6, 3, 9, 2, 8, 11],
+    [13, 11, 7, 14, 12, 1, 3, 9, 5, 0, 15, 4, 8, 6, 2, 10],
+    [6, 15, 14, 9, 11, 3, 0, 8, 12, 2, 13, 7, 1, 4, 10, 5],
+    [10, 2, 8, 4, 7, 6, 1, 5, 15, 11, 9, 14, 3, 12, 13, 0],
+];
+
+pub const XOR_VALUES: [u64; XOR_4BITS_NUMBER] = [
+    0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 1, 0, 3, 2, 5, 4, 7, 6, 9, 8, 11, 10, 13,
+    12, 15, 14, 2, 3, 0, 1, 6, 7, 4, 5, 10, 11, 8, 9, 14, 15, 12, 13, 3, 2, 1, 0, 7, 6, 5, 4, 11,
+    10, 9, 8, 15, 14, 13, 12, 4, 5, 6, 7, 0, 1, 2, 3, 12, 13, 14, 15, 8, 9, 10, 11, 5, 4, 7, 6, 1,
+    0, 3, 2, 13, 12, 15, 14, 9, 8, 11, 10, 6, 7, 4, 5, 2, 3, 0, 1, 14, 15, 12, 13, 10, 11, 8, 9, 7,
+    6, 5, 4, 3, 2, 1, 0, 15, 14, 13, 12, 11, 10, 9, 8, 8, 9, 10, 11, 12, 13, 14, 15, 0, 1, 2, 3, 4,
+    5, 6, 7, 9, 8, 11, 10, 13, 12, 15, 14, 1, 0, 3, 2, 5, 4, 7, 6, 10, 11, 8, 9, 14, 15, 12, 13, 2,
+    3, 0, 1, 6, 7, 4, 5, 11, 10, 9, 8, 15, 14, 13, 12, 3, 2, 1, 0, 7, 6, 5, 4, 12, 13, 14, 15, 8,
+    9, 10, 11, 4, 5, 6, 7, 0, 1, 2, 3, 13, 12, 15, 14, 9, 8, 11, 10, 5, 4, 7, 6, 1, 0, 3, 2, 14,
+    15, 12, 13, 10, 11, 8, 9, 6, 7, 4, 5, 2, 3, 0, 1, 15, 14, 13, 12, 11, 10, 9, 8, 7, 6, 5, 4, 3,
+    2, 1, 0,
+];
+
+fn split_value_4bits<F: PrimeField + Hash>(value: u128, n: usize) -> Vec<F> {
```

Comment:
You can make a parameter mutable, `(mut value: u128` so it is not necessary the first line

Reply:
It is because this parameter maybe a sum of two or three values.
For example:
```rust 
va_bit_vec.push(split_value_4bits(
    v1_vec_values[a] as u128 + v1_vec_values[b] as u128 + x as u128,
    BITS_NUMBER + 1,
));
```
So if we make the parameter mutable, we need to defined a new mutable value outside before we call the split_value_4bits function. Code readability doesn't seem to be better.

Reply:
I don't think that is the case. u128 implements the Copy trait. So you can pass a immutatble variable to a mutable parameter, because it is always a copy in the parameter.

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=341b4f4b4eb26ca6fb409f2052349ea4